### PR TITLE
KAFKA-15853: Move KafkaConfig log properties and docs out of core

### PR DESCRIPTION
**Pull request by OmniaGM**
part 1 for #15501
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by OmniaGM** [comment]
@nizhikov can you have a look please?

File: server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java
Comment (nizhikov) [inline server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java]: Minor 

1. Let's rename this to `LOG_PREFIX` to make naming more compact.

2. We have `ServerTopicConfigSynonyms.LOG_PREFIX`, already.
Maybe it will be better to use only one constant?

**Review by nizhikov — APPROVED** [review]

File: build.gradle
Comment (nizhikov) [inline build.gradle]: Tests not required in `storage` module - only main classes. Line 1841 can be removed.

File: server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java
Comment (OmniaGM) [reply server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java]: Nice catch. I updated the pr to `ServerTopicConfigSynonyms.LOG_PREFIX`

File: server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java
Comment (mimaison) [inline server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java]: For consistency with the other *Configs classes, should we used the `CONFIG` suffix instead of `PROP`?

Comment (mimaison) [reply server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java]: Would it also make sense to colocate the `CONFIG` and `DOC` like we do in the other Config classes?

File: core/src/main/scala/kafka/server/ZkAdminManager.scala
Comment (mimaison) [inline core/src/main/scala/kafka/server/ZkAdminManager.scala]: I wonder if we should just import `KafkaConfig` and use `KAFKA.SOME_CONFIG` instead of importing all/statically KafkaConfig. WDYT?

**Review by mimaison — COMMENTED** [review]
Thanks for the PR. I left a few suggestions.

File: server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java
Comment (OmniaGM) [reply server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java]: > For consistency with the other *Configs classes, should we used the CONFIG suffix instead of PROP?

I thought about it but I wan't sure which would make it easier for people to review. 
We also might need to do the same for LogCleaner configs as well. 

> Would it also make sense to colocate the CONFIG and DOC like we do in the other Config classes?

I considered this but I found it tricky for some docs that refer to more than 1 config in the string interpolation.  For example the `LOG_DIR_DOC` and `LOG_DIRS_DOC` refer each others prop e.g LOG_DIRS_DOC  refer to `LOG_DIR_PROP` and LOG_DIR_DOC  refers to `LOG_DIRS_PROP` same goes to `LOG_ROLL_TIME_HOURS_DOC` and `LOG_ROLL_TIME_MILLIS_DOC`. I thought this why they are grouped at the end after defining the properties names first.

Comment (OmniaGM) [reply server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java]: I just pushed a commit to replace `PROP` with `CONFIG` suffix.

File: core/src/main/scala/kafka/server/ZkAdminManager.scala
Comment (OmniaGM) [reply core/src/main/scala/kafka/server/ZkAdminManager.scala]: <del> For this one and maybe other that import few config from KafkaConfig this would be nice improvement but some classes like DynamicBrokerConfig for example include a large set of config. I can update few to use  `KAFKA.SOME_CONFIG` if they need small set of config and leave the others specially that we when we move them to java we will do this anyway.</del>

Ignore my previous comment I got the comment wrong. At the moment we can't use KafkaConfig everywhere as there's another KafkaConfig in scala

Comment (OmniaGM) [reply core/src/main/scala/kafka/server/ZkAdminManager.scala]: Also at the moment we can't use `KafkaConfig` everywhere as there's another `KafkaConfig` in scala

Comment (mimaison) [reply core/src/main/scala/kafka/server/ZkAdminManager.scala]: Ah right, that's a good point. Thanks

**Comment by mimaison** [comment]
Now that https://github.com/apache/kafka/pull/15075 got merged, this needs rebasing.

**Comment by OmniaGM** [comment]
rebased the pr

File: server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java
Comment (mimaison) [inline server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java]: Should it be `Documentation`?

File: server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java
Comment (mimaison) [inline server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java]: Can we use a consistent formatting for down conversion across classes? Either `DOWN_CONVERSION` or `DOWNCONVERSION` but let's pick one.

**Review by mimaison — COMMENTED** [review]
Thanks for the updates! I made another pass and left a couple of comments.

**Review by fvaleri — APPROVED** [review]
LGTM, thanks.

**Comment by showuon** [comment]
Since Mickael is already reviewing it, I'll focus on other PRs. Thanks @mimaison !

**Comment by fvaleri** [comment]
@OmniaGM I think this is almost ready, but there are a couple of minor comments to address and some conflicts to fix.

**Comment by OmniaGM** [comment]
> @OmniaGM I think this is almost ready, but there are a couple of minor comments to address and some conflicts to fix.

Will be addressing this soon after another pr

**Comment by OmniaGM** [comment]
@fvaleri, @nizhikov @mimaison just rebased the pr and addressed the feedback will wait for this one to be merged before rebasing the other pr for KafkaConfig

**Comment by mimaison** [comment]
@OmniaGM Is this the next PR in line for moving broker configs?

**Comment by OmniaGM** [comment]
> @OmniaGM Is this the next PR in line for moving broker configs?

Yes, I rebased the PR and also pushed a refactor to follow the same pattern we adopted in https://github.com/apache/kafka/pull/15656#discussion_r1554518847, https://github.com/apache/kafka/pull/15684, https://github.com/apache/kafka/pull/15670. Now the server configs for logs are in central class called KafkaLogConfigs and this will be used between broker and any other module instead of having a big KafkaConfig with everything. We can relay on `org.apache.kafka.server.config` to be the package where all configs are defined/listed for broker side

cc:@chia7712

File: core/src/main/scala/kafka/server/ZkAdminManager.scala
Comment (OmniaGM) [reply core/src/main/scala/kafka/server/ZkAdminManager.scala]: After the last commit, this shouldn't be an issue anymore as the configs are in KafkaLogConfigs instead of KafkaConfig so there is no conflict anymore

File: server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java
Comment (OmniaGM) [reply server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java]: I reorganised this KafkaLogConfigs in the last commit to have config, doc and default in the same order like the other config classes.

**Comment by OmniaGM** [comment]
The last commit renamed the class to `KafkaLogConfigs` instead; similar to `KafkaSecurityConfigs` and `ZKConfigs`. I tried to move some to `LogConfig.ServerLogConfig` similar to `LogConfig.RemoteLogConfig` specially that some default configs are defined in `LogConfig`. However I'm bit concern that `LogConfig` seems already huge. What others prefer? Keep it in  `KafkaLogConfigs`  or move them to `LogConfig.ServerLogConfig`
cc: @mimaison, @chia7712

**Comment by chia7712** [comment]
> Keep it in KafkaLogConfigs

I prefer this way since `LogConfig` includes too many stuff now ... maybe we should move the constants out of `LogConfig`

File: server/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java
Comment (chia7712) [inline server/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java]: Should we align the declaration style? I prefer the style: `static final`

**Comment by OmniaGM** [comment]
> maybe we should move the constants out of `LogConfig`

We can try to do this in a follow up after we merge all PRs for KafkaConfig as am worried about potential circular dependency. And it would better to see the full pic before do improvement for `LogConfig`.

**Comment by chia7712** [comment]
There are build error. please fix it :)

File: server/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java
Comment (chia7712) [inline server/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java]: It seems to me we should make consistency for all docs here. 1) make it point to `TopicConfig.xxxDOC` or 2) copy the docs to here.

**Review by chia7712 — COMMENTED** [review]
@OmniaGM thanks for updated PR. overall +1 and please take a look at one last comment

File: server/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java
Comment (OmniaGM) [reply server/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java]: I don't think we can copy all configs doc to `TopicConfig` as the docs are different in many places between topic level config and log/topic broker level config. I will copy it for now into KafkaLogConfigs in case the doc diverted in the futrue like the rest of the docs KafkaLogConfigs.

**Comment by fvaleri** [comment]
@OmniaGM please rebase, hopefully for the last time :)

**Comment by chia7712** [comment]
> hopefully for the last time :)

I feel #15684 will cause many conflicts to this PR, and vice versa :_

**Comment by mimaison** [comment]
Let's just let Omnia decides which PR she prefers doing first. Then we can all review that PR, and once merged move onto the next one. This would avoid unnecessary churn as I expect she's getting tired of rebasing all of these PRs every day!

**Comment by OmniaGM** [comment]
> Let's just let Omnia decides which PR she prefers doing first. Then we can all review that PR, and once merged move onto the next one. This would avoid unnecessary churn as I expect she's getting tired of rebasing all of these PRs every day!

Thanks @mimaison this is really thoughtful! breaking KafkaConfig out of core is one of the hardest cleanup specially if we want to move with smaller/medium PRs. So I don't see any other way than going through it. I think it might be better to take them in order so we can get them merged and the rebases can be easier. 
Let's merge https://github.com/apache/kafka/pull/15684 as
-  it is smaller 
-  more contained than the log config 
- and group configs are more likely to change as part of KIP-848 and any of its followups which make the conflicts much harder and risker to get messed up. It's kinda easier to deal with conflict caused by other KafkaConfig cleanups than conflicts caused by other KIPs, Jiras implementations 
- it also has bit of more priority for https://issues.apache.org/jira/browse/KAFKA-15089.

File: core/src/main/scala/kafka/server/ZkAdminManager.scala
Comment (mimaison) [inline core/src/main/scala/kafka/server/ZkAdminManager.scala]: This should be `CREATE_TOPIC_POLICY_CLASS_NAME_CONFIG`

File: core/src/test/scala/unit/kafka/log/LogConfigTest.scala
Comment (mimaison) [inline core/src/test/scala/unit/kafka/log/LogConfigTest.scala]: `LOG_MESSAGE_TIMESTAMP_DIFFERENCE_MAX_MS_PROP` has been renamed to `LOG_MESSAGE_TIMESTAMP_BEFORE_MAX_MS_CONFIG`
Same below

File: core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala
Comment (mimaison) [inline core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala]: This has been renamed to `LOG_FLUSH_INTERVAL_MS_CONFIG`

**Review by mimaison — COMMENTED** [review]
Thanks for the updates. I found a few small issues.

File: core/src/main/scala/kafka/server/ZkAdminManager.scala
Comment (mimaison) [inline core/src/main/scala/kafka/server/ZkAdminManager.scala]: This one was correct! We need to undo this

Comment (OmniaGM) [reply core/src/main/scala/kafka/server/ZkAdminManager.scala]: fix it

File: server/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java
Comment (mimaison) [inline server/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java]: Nit: There's a mix of `DOWNCONVERSION` and `DOWN_CONVERSION`.

**Review by mimaison — APPROVED** [review]
Thanks for all the updates and quickly addressing feedback. LGTM

File: server/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java
Comment (OmniaGM) [reply server/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java]: Good spot. fixed now

**Comment by chia7712** [comment]
> However I'm bit concern that LogConfig seems already huge. What others prefer? Keep it in KafkaLogConfigs or move them to LogConfig.ServerLogConfig

the most default values of `KafkaLogConfigs` are in `LogConfig`, and they are in different module. That pattern is no similar to `ReplicationConfigs`, `KafkaSecurityConfigs`. Is server-common module more suitable to collect those server side configs since both`storage` and `server` depend on `server-common`. Also, `server-common` has `org.apache.kafka.server.config` package too.

**Comment by OmniaGM** [comment]
> > However I'm bit concern that LogConfig seems already huge. What others prefer? Keep it in KafkaLogConfigs or move them to LogConfig.ServerLogConfig
> 
> the most default values of `KafkaLogConfigs` are in `LogConfig`, and they are in different module. That pattern is no similar to `ReplicationConfigs`, `KafkaSecurityConfigs`. Is server-common module more suitable to collect those server side configs since both`storage` and `server` depend on `server-common`. Also, `server-common` has `org.apache.kafka.server.config` package too.

I think it might be better and simpler to move `KafkaLogConfigs` into `server-common` and make LogConfig use them. I'll push an update with this soon

File: server-common/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java
Comment (chia7712) [inline server-common/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java]: Do we need to add prefix "kafka" here? sorry that I try to figure out the naming rule. It seems to me the prefix "kafka" means the class is used to collect other "xxxConfigs" constants and to be the central place for kafka xxx-related configs. For example, `KafkaSecurityConfigs` collects all security-related configs from `BrokerSecurityConfigs`, `SaslConfigs`, `SslConfigs`, etc

**Review by chia7712 — COMMENTED** [review]
@OmniaGM thanks for updated PR. overall +1. one last comment left.

File: server-common/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java
Comment (OmniaGM) [reply server-common/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java]: I can rename it to `ServerLogConfigs` or `BrokerLogConfigs` but we already have `LogConfig`

Comment (OmniaGM) [reply server-common/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java]: Just pushed an update. The problem right now that we have the name of the config type without any unnecessary prefixes has been preserved in other modules and am trying to avoid making this pr bigger than it should be. I think at some point we need to revisit some of the existing Config files as some of them breaks few rules like using `PROP` instead of `CONFIG` suffix and use `DEFAULT` as prefix instead of suffix and some of the classes is `Config` instead of `Configs` as well.

Comment (chia7712) [reply server-common/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java]: > I think at some point we need to revisit some of the existing Config files as some of them breaks few rules like using PROP instead of CONFIG suffix and use DEFAULT as prefix instead of suffix and some of the classes is Config instead of Configs as well.

agree!

**Review by chia7712 — APPROVED** [review]
LGTM, let us move on

**Comment by chia7712** [comment]
```
[2024-04-18T17:26:16.149Z] [ant:checkstyle] [ERROR] /home/jenkins/workspace/Kafka_kafka-pr_PR-15569/server-common/src/main/java/org/apache/kafka/server/config/ServerLogConfigs.java:22:1: Disallowed import - org.apache.kafka.server.common.MetadataVersion.IBP_3_0_IV1. [ImportControl]

[2024-04-18T17:26:16.149Z] [ant:checkstyle] [ERROR] /home/jenkins/workspace/Kafka_kafka-pr_PR-15569/server-common/src/main/java/org/apache/kafka/server/config/ServerLogConfigs.java:23:1: Disallowed import - org.apache.kafka.server.config.ServerTopicConfigSynonyms.LOG_PREFIX. [ImportControl]

```
there are imports error :_

**Comment by chia7712** [comment]
The failed test `testParseAndValidateAddressesWithReverseLookup` will get fixed by #15758. I will merge this PR

**Pull request merged**
